### PR TITLE
Feature/generic service provider flow closes #44 and #43

### LINKED ISSUE #44: 500 Error /hello endpoint [High]
**Bug Summary**

Getting 500 server error when hitting /hello (currently mapped to reddit flow).
See screenshot below.

**Test Case(s)**
N/A (we should have integration tests for this)

---

**Expected behavior**

"hello" string returned as JSON object.
Logger outputs reddit posts.

**Observed Behaviour**

500 server error

**Steps To Reproduce**

1. Populate application.properties with reddit credentials
2. Start spring server
3. Make a GET to /hello 

**Environment**
- Version: latest
- OS: macos (latest)
- Browser: chrome (latest)

---

**Code Examples**

**Stack Trace**
```
java.lang.NullPointerException: null
	at com.feeder.server.providers.reddit.RedditFeedProvider.getFeed(RedditFeedProvider.java:61) ~[classes/:na]
	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
Error has been observed at the following site(s):
	|_ checkpoint ⇢ HTTP GET "/hello" [ExceptionHandlingWebHandler]
Stack trace:
		at com.feeder.server.providers.reddit.RedditFeedProvider.getFeed(RedditFeedProvider.java:61) ~[classes/:na]
		at com.feeder.server.controllers.HelloController.readHello(HelloController.java:28) ~[classes/:na]
		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
		at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
		at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
		at org.springframework.web.reactive.result.method.InvocableHandlerMethod.lambda$invoke$0(InvocableHandlerMethod.java:147) ~[spring-webflux-5.2.4.RELEASE.jar:5.2.4.RELEASE]
		at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:151) ~[reactor-core-3.3.3.RELEASE.jar:3.3.3.RELEASE]
		at reactor.core.publisher.MonoFlatMap.subscribeOrReturn(MonoFlatMap.java:53) ~[reactor-core-3.3.3.RELEASE.jar:3.3.3.RELEASE]
		at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:48) ~[reactor-core-3.3.3.RELEASE.jar:3.3.3.RELEASE]
		at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52) ~[reactor-core-3.3.3.RELEASE.jar:3.3.3.RELEASE]
		at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.drain(MonoIgnoreThen.java:153) ~[reactor-core-3.3.3.RELEASE.jar:3.3.3.RELEASE]
		at reactor.core.publisher.MonoIgnoreThen.subscribe(MonoIgnoreThen.java:56) ~[reactor-core-3.3.3.RELEASE.jar:3.3.3.RELEASE]
		at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:55) ~[reactor-core-3.3.3.RELEASE.jar:3.3.3.RELEASE]
		at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:150) ~[reactor-core-3.3.3.RELEASE.jar:3.3.3.RELEASE]
		at reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onNext(FluxSwitchIfEmpty.java:67) ~[reactor-core-3.3.3.RELEASE.jar:3.3.3.RELEASE]
		at reactor.core.publisher.MonoNext$NextSubscriber.onNext(MonoNext.java:76) ~[reactor-core-3.3.3.RELEASE.jar:3.3.3.RELEASE]
...
```
**Screenshots**
image

**Error Report**

---

**Additional context**

### PR DESCRIPTION
**Pull request by harmanlamba**
GitHub Issue (If applicable): # 43 and 44

https://github.com/softeng-701-group-5/softeng-701-assignment-1/issues/43
https://github.com/softeng-701-group-5/softeng-701-assignment-1/issues/44

## PR Type

What kind of change does this PR introduce?

- Bugfix
- Feature

## What is the current behavior?

There is no generic service provider interface of workflow for the API teams to adhere to. Additionally, there was a bug with the RedditServiceProvider which gave a 500 error when trying to access /hello. 

## What is the new behavior?

A new generic service provider interface of workflow has been created for the API teams to adhere to. Skeleton templates have been created, which can be modified by each API team accordingly. Furthermore, issue #44's bug has been fixed and patched in this PR.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [X] Tested code with current upstream repo
- [X] Added new tests to the test suite for the feature
- [X] Tested the full test suite, and ensured that the feature does not break current build.
- [ ] Docs have been added/updated which fit (for bug fixes / features)
- [X] Contains **NO** breaking changes
- [X] Associated with an issue (GitHub or internal)
- [X] Checked licensing of Frameworks/Libraries (if applicable) 

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: server/src/main/java/com/feeder/server/model/DemoData.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+@JsonDeserialize(builder = AutoValue_DemoData.Builder.class)
+public abstract class DemoData extends GenericData {
+
+  public static Builder newBuilder() {
+    return new AutoValue_DemoData.Builder();
+  }
+
+  @Override
+  public Type feedType() {
+    return Type.DEMO;
```
Comment (jordansimsmith) [inline server/src/main/java/com/feeder/server/model/DemoData.java]: Does this serialise as an integer or as a string? Ideally, the frontend would deal with strings (i.e `type === 'github') as the enum <> int mapping is not exposed.

Comment (Minus20Five) [reply server/src/main/java/com/feeder/server/model/DemoData.java]: Its a String.

**Comment by harmanlamba** [comment]
> Awesome job guys 🎉, looks pretty good to me. As @jordansimsmith said, do we need unit tests for the `serializationDemoFlow()`?
> 
> As mentioned in the [brief](https://canvas.auckland.ac.nz/courses/47939/files/folder/Theme%201/coursework?preview=4047131):
> 
> > All  additions  or  modifications  to  code  should  include  associated  tests

Hi @HarrisonLeach1, you make a good point. I'll add some code that tests the controller. I'll try to get it done by tonight, if not have it done by tomorrow.

**Comment by jordansimsmith** [comment]
Added do not merge label until tests have been added 👍

**Comment by Minus20Five** [comment]
> ... @jordansimsmith said, do we need unit tests for the `serializationDemoFlow()`?

Integration test has been added. For now, all the dependencies (i.e. `FeedProvider`) have been mocked.

> Regarding the 2 width tabs, is this chosen as part of #13? We should be consistent either way

Yes, I'm probably going with Google convention.

Re-requesting approval 🙂.

**Comment by jordansimsmith** [comment]
@Minus20Five LGTM. Some of the new commits have 4 space tabs, maybe these should be fixed but no big deal.

**Comment by Minus20Five** [comment]
> Some of the new commits have 4 space

I think for wrapping its 4 spaces.

**Pull request merged**